---
title: "Class 13: Genome informatics 1"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## Interested in abundance SNPs of asthma in individiuals of mexican ancestry in Los Angelos.

Read in our data from omim:

```{r}
asthma_mxl <- read.csv("373531-SampleGenotypes-Homo_sapiens_Variation_Sample_rs8067378 (1).csv")
asthma_mxl
```

If we want the abundances of the variations:
```{r}
table(asthma_mxl$Genotype..forward.strand.) / nrow(asthma_mxl) * 100
```

## FASTQ Quality Scores

```{r}
#Install our packages and open them
#install.packages("seqinr")
#install.packages("gtools")

library(seqinr)
library(gtools)
chars <- s2c("DDDDCDEDCDDDDBBDDDCC@")
chars
phred <- asc(chars) - 33
phred
```

```{r}
samples <- read.table("rs8067378_ENSG00000172057.6.txt")
samples
summary(samples)
table(samples$geno)
inds.gg <- samples$geno == "G/G"
inds.ag <- samples$geno == "A/G"
inds.aa <- samples$geno == "A/A"
summary(samples[inds.gg,"exp"])
box.gg <- samples[inds.gg,"exp"]
box.ag <- samples[inds.ag,"exp"]
box.aa <- samples[inds.aa,"exp"]
boxplot(exp ~ geno, data=samples, notch=TRUE)

```

